### ISSUE TITLE
Can't add ManytoMany fields to a Page

### ISSUE THREAD

**Issue by bufke** [wagtail/wagtail#231]
For example

```
class BlogPage(Page):
    body = RichTextField()
    categories = models.ManyToManyField(BlogCategory, blank=True)
```

One gets ValueError "<BlogPage: foo>" needs to have a value for field "blogpage" before this many-to-many relationship can be used.

It's caused by this 
https://github.com/torchbox/django-modelcluster/issues/3
https://github.com/torchbox/django-modelcluster/blob/master/modelcluster/forms.py#L260

I'll keep trying to think of a solution. Seems like a common use case as Wordpress implements categories as a M2M (you can pick multiple, but they are predefined unlike a tag).

**Comment by bufke** [comment]
I'm trying to determine if this is something I could fix. The main issue is M2M support in modelcluster. What's the scope of fixing something like that? I have experience with making django apps but it's not something I could commit weeks too either.

Another small component would be a UI widget but that seems straightforward.

**Comment by mhnbcu** [comment]
:+1: for a fix on this. 

Agree with @mattots that tags don't quite negate the need for M2M.

**Comment by gasman** [comment]
@bufke Re the scope of this fix - unfortunately it's the sort of problem where figuring out the scope is 90% of the battle, so not really the sort of thing that I can give any pointers on. It would certainly involve digging into Django internals.

**Comment by bufke** [comment]
@gasman thanks I started reading through the modelcluster source. M2M would be a significant project for sure. I'm questioning the logic in supporting it actually. A M2M inherently affects multiple objects (pages perhaps). Would I ever want to edit/create M2M objects in the cluster way of keeping them in ram till saving to the database? I don't think I would.

Let's consider how wordpress does it - If I create a post, assign it to a new category, then save as a draft - the category is created and visible on other pages. That's probably fine. 

Perhaps this would be best solved with a special M2M widget so the workaround doesn't look hacky. It's a pretty awkward UX right now due to using InlinePanel in an odd manner.

![screenshot from 2015-03-11 14 59 21](https://cloud.githubusercontent.com/assets/739307/6604536/4e547120-c7ff-11e4-9575-2c6675bd1325.png)

What if we developed something like the django admin many to many widget. Maybe ManyToManyPanel. Then document how to use ManyToMany and note you must use a through table as to specify the ParentalKey. Finally I'd propose the documentation suggest making Category a snippet so the user can actually edit it.

If I can whip this idea together I'd be happy to submit upstream.

**Comment by gasman** [comment]
@bufke I don't think we'd ever want to make the objects at the other end of an M2M relation editable as part of the page - that would cause all sorts of headaches if multiple pages had different draft edits of the same objects. The overall functionality would remain the same as it is with the workaround: being able to specify objects from an existing master list (defined elsewhere - e.g. in snippets) to be associated with this page. (We could still support adding / editing items without leaving the page editor, in the same sense that we currently do for images.)

The benefits of supporting M2M relations 'properly' would be:
* a more suitable UI widget
* a neater and more compact representation when serialized to JSON (e.g. when saving a page revision) - we'd store it as a list of IDs, rather than a list of through models that exist solely to hang an ID off.
* Not having to design your database schema around Wagtail's requirements. An important design goal for Wagtail is that you should be able to design your schema in the way that best fits your data, not have it dictated by Wagtail's own implementation details.

Right now we have no ETA for getting that full solution implemented, so I'd be very happy to consider an approach that avoids the UI weirdness, as an interim solution. Thanks!

**Comment by benje** [comment]
thanks for the workaround @bufke

**Comment by bufke** [comment]
I'd be interested in starting a bounty on this issue. Maybe use bounty source? I could throw in $1000 for a good fix that got proper M2M support in modelcluster and a widget for wagtail.

**Comment by gasman** [comment]
@bufke There's a PR in progress at https://github.com/torchbox/django-modelcluster/pull/36 - if there were to be a bounty then it would be fair for it go to @theju (assuming it works as advertised, of course). Are retrospective bounties a thing...? :-)

It's currently waiting on Django 1.7 support, but at this point I reckon Django 1.7 is far enough in the past that we could reasonably ship this as a 1.8-only feature (provided that it doesn't break any existing 1.7 functionality). In that case, the only thing blocking this is me finding the time to review it all, since it's quite a complex update - I'll make this a priority for Wagtail 1.3.

**Comment by ababic** [comment]
Hi guys. I know it's difficult to know exactly when this will be sorted, due to the modelcluster dependency, but I wonder if it would be worth some work in the meantime, to try and at least solve some of the editing limitations that the lack of ManyToManyField presents?

For example, where you very simple `Category` model, and have already created a 'through' model to associate your page with categories, would it be super difficult to offer an alternative to `InlinePanel`, that would provide a checkbox list for you to select from? Or perhaps something more complicated (django's filter_horizontal widget) if there was a need for it.

Maybe it wouldn't be worth the work if 1.4 is coming soon?

**Comment by gasman** [comment]
Hi @ababic,

A nice form field for M2M relations is something that we're going to want either way, and up to now I'd anticipated that it would be the second stage of development, after the modelcluster updates. However, if you think it's something that could usefully be done now, using 'through' models as an interim solution, then that would certainly be welcome.

On the other hand, it may be that with modelcluster's M2M support in place, the UI work simply reduces to "use ModelMultipleChoiceField"... in which case it probably is better to wait for modelcluster. That task is gradually rising to the top of my todo list, and will hopefully get done in the next two weeks or so.

**Comment by davejacobs** [comment]
That's great to hear! Can't wait for this.

**Comment by SableWalnut** [comment]
:+1:

**Comment by melinath** [comment]
Just ran into this as well. The current M2M widget should work well enough for now, but it would be nice to have an autosuggest field instead. I'm a little concerned about the idea of requiring everything to be modelcluster-compatible to have this working properly :-/

**Comment by kofic** [comment]
@gasman Is this available now?

**Comment by gasman** [comment]
@kofic No (but I'm working on it) - there'll be an update here when there's any news.

**Comment by gasman** [comment]
This is now implemented in django-modelcluster: https://github.com/torchbox/django-modelcluster/pull/62

From initial tests, it seems to be working nicely with Wagtail: just use ParentalManyToManyField instead of ManyToManyField, and specify it in a FieldPanel as normal.

Things that need to be done before the release of Wagtail 1.9:

* [ ] release django-modelcluster 3.0 final and update wagtail's setup.py to require `django-modelcluster>=3.0,<4.0`
* [ ] add a release note to document the updated django-modelcluster / django-taggit dependencies (since it's somewhat likely that people will have the old versions hard-coded in their requirements files)
* [ ] ensure that M2M relations are covered by tests
* [ ] document the use of ParentalManyToManyField, possibly through a new 'categories' section in the tutorial (which was dropped from https://github.com/wagtail/wagtail/pull/3125#issuecomment-259181336 because it was too messy to implement without real M2M support) as well as http://docs.wagtail.io/en/v1.8/topics/pages.html

**Comment by nickhudkins** [comment]
Before opening another ticket, which I believe would be a duplicate of: https://github.com/wagtail/wagtail/issues/2635, does the work being done here apply only to ParentalManyToManyField or will we be able to utilize this in `contrib.modeladmin` as well, for standard M2Ms and ForeignKeys. I understand that I could also just roll my own here, the use case being, a ForeignKey to a model with 100k rows destroys a select box. In the standard django-admin I have always used: https://github.com/lincolnloop/django-salmonella to get around this sort of problem. 

Thank you for all your work!

**Comment by gasman** [comment]
Hi @nickhudkins - in retrospect I think I probably misinterpreted #2635. Since the only way to achieve M2M-like relations on pages was with 'through' models in an InlinePanel (it hadn't occurred to me that ManyToManyField does work on snippet / modeladmin models), I took it to be a request for a "better widget" than the through-model / InlinePanel setup.

Either way, the resolution is the same - now that we consistently support 'proper' M2Ms everywhere, the process for swapping in an alternative widget via `FieldPanel` is the same as for any other field type - and, as you observe, coming up with a better widget isn't a Wagtail-specific issue.

**Comment by kofic** [comment]
This is awesome @gasman I will test it out

**Comment by gasman** [comment]
Completed in #3305.
